Improve handling of polyfilled assets (and improve handling of Reader themes data)

johnwatkins0 opened the pull request:
## Summary

This was initially intended to address #5147. I wasn't able to reproduce that issue but I took a couple of steps to improve polyfilling in hopes that the issue will be fixed.

1. Removed `lodash ` from bundled assets. It shouldn't have been there in the first place; this was an oversight. By loading the version provided by core and setting the `_.noConflict` inline script I'm hoping the issue brought up in #5147 will be fixed.
2. Created a `Polyfills` class that registers core assets that may not be available in earlier versions of WP.
3. Created a new action hook that is placed throughout the plugin where the polyfilled core assets may need to be registered. the new service is delayed until this action, effectively meaning it will not be registered on screens where the action hook is never called.

<!-- Please reference the issue this PR addresses. -->
Fixes #5147

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

johnwatkins0 committed:
Improve handling of polyfilled assets

johnwatkins0 committed:
Move script polyfills from amp_register_default_scripts to Polyfills class

johnwatkins0 committed:
Remove lodash from bundles (shouldn't have been included in the first place)

johnwatkins0 commented on src/Admin/RESTPreloader.php:
```diff
@@ -43,6 +43,10 @@ public function add_preloaded_path( $path ) {
 	 * Preloads data using apiFetch preloading middleware.
 	 */
 	public function preload_data() {
+		if ( ! function_exists( 'rest_preload_api_request' ) ) { // Not available pre-5.0.
```
Caught this after switching to WP 4.9 locally.

johnwatkins0 committed:
Initiate scripts and styles in tests

johnwatkins0 committed:
Polyfill assets during paired browsing

johnwatkins0 committed:
Modify conditional check for class registration

johnwatkins0 committed:
Remove conditional and delayed interfaces from polyfills class

johnwatkins0 committed:
Polyfill assets during paired browsing

johnwatkins0 commented on src/Admin/Polyfills.php:
```diff
@@ -0,0 +1,130 @@
+ *
+ * @since 2.0
+ */
+final class Polyfills implements Service, Registerable {
```
I made an effort to make this class `Delayed` and `Conditional` but there were several moving parts involving both the admin and paired mode (and the iframes within paired mode), and some paired-mode unit tests were failing. I could tell it was going to take me a while to work through it, so instead, for now, we're running this class all the time. There are still conditionals, though, to only register assets if they are not already registered in core. Open to suggestions for handling this better.

pierlon replied:
I'd say this is one the few services where it needs to always be loaded. As for delaying it, I'm not sure which action would be most appropriate to use; I'd defer to @westonruter.

schlessera replied:
If the `Conditional` is too complex to use from here, it could potentially be turned around. We could say it'd only register on `'amp_register_polyfills'` and then `do_action( 'amp_register_polyfills' )`  in the parts of the code where we need them.

Not making this Conditional would mean we load it for other plugins as well, and could for example break if another plugin hard-codes shims on its own settings page.

johnwatkins0 replied:
Thanks for this idea. Done in c955d3a8e8edd0118202b994dc1ff4561719d7d2

pierlon commented on tests/php/src/Admin/PolyfillsTest.php:
```diff
@@ -0,0 +1,79 @@
+ *
+ * @since 2.0
+ *
+ * @covers Polyflls.
```
```suggestion
 * @covers Polyfills.
```

pierlon commented on tests/php/src/Admin/PolyfillsTest.php:
```diff
@@ -0,0 +1,79 @@
+		$this->assertTrue( wp_style_is( 'wp-components', 'registered' ) );
+
+		unset( $wp_scripts );
+		unset( $wp_styles );
```
```suggestion
		unset( $wp_scripts, $wp_styles );
```

schlessera commented on src/Admin/Polyfills.php:
```diff
@@ -0,0 +1,130 @@
+
+				$wp_scripts->add(
+					$handle,
+					amp_get_asset_url( sprintf( 'js/%s.js', $handle ) ),
```
No need for an expensive `sprintf()` here:
```suggestion
					amp_get_asset_url( "js/{$handle}.js" ),
```

schlessera commented on src/Admin/Polyfills.php:
```diff
@@ -0,0 +1,130 @@
+		$handles = [ 'wp-i18n', 'wp-dom-ready', 'wp-polyfill', 'wp-url' ];
+		foreach ( $handles as $handle ) {
+			if ( ! isset( $wp_scripts->registered[ $handle ] ) ) {
+				$asset_file   = AMP__DIR__ . '/assets/js/' . $handle . '.asset.php';
```
We're defaulting to curly brace substitution in PHP:
```suggestion
				$asset_file   = AMP__DIR__ . "/assets/js/{$handle}.asset.php";
```

schlessera commented on src/Admin/Polyfills.php:
```diff
@@ -0,0 +1,130 @@
+
+		if ( ! isset( $wp_scripts->registered['wp-api-fetch'] ) ) {
+			$asset_handle = 'wp-api-fetch';
+			$asset_file   = AMP__DIR__ . '/assets/js/' . $asset_handle . '.asset.php';
```
```suggestion
			$asset_file   = AMP__DIR__ . "/assets/js/{$asset_handle}.asset.php";
```

schlessera commented on src/Admin/Polyfills.php:
```diff
@@ -0,0 +1,130 @@
+
+			$wp_scripts->add(
+				$asset_handle,
+				amp_get_asset_url( 'js/' . $asset_handle . '.js' ),
```
```suggestion
				amp_get_asset_url( "js/{$asset_handle}.js" ),
```

schlessera commented on src/Admin/Polyfills.php:
```diff
@@ -0,0 +1,130 @@
+		/*
+		* Polyfill dependencies that are registered in Gutenberg and WordPress 5.0.
+		* Note that Gutenberg will override these at wp_enqueue_scripts if it is active.
+		*/
```
```suggestion
		/*
		 * Polyfill dependencies that are registered in Gutenberg and WordPress 5.0.
		 * Note that Gutenberg will override these at wp_enqueue_scripts if it is active.
		 */
```

johnwatkins0 committed:
Comment spacing fix

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

johnwatkins0 committed:
Use curly-brace substitution instead of sprintf/concatenation

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>
Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Add action hook for registration of polyfills

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into fix/5147-missing-underscore-find-function

johnwatkins0 commented on webpack.config.js:
```diff
@@ -229,6 +229,7 @@ const setup = {
 			// Most dependencies will be bundled for the AMP setup screen for compatibility across WP versions.
 			requestToHandle: ( handle ) => {
 				switch ( handle ) {
+					case 'lodash':
```
Note: if these changes don't fix the issue in #5147 then we may instead want to emulate the approach in Site kit here: https://github.com/google/site-kit-wp/pull/1153/files
This would mean going back to bundling lodash into our JS.

pierlon reviewed (changes requested):
`\AMP_Template_Customizer::enqueue_legacy_preview_scripts`, `\AMP_Template_Customizer::add_legacy_customizer_scripts` and `\AMP_Template_Customizer::add_customizer_scripts` would need to execute the `amp_register_polyfills` action as well.

Also, one of those `do_action` calls should have a docblock.

johnwatkins0 committed:
Call amp_register_polyfills action in a few more places and document it

johnwatkins0 commented:
> `\AMP_Template_Customizer::enqueue_legacy_preview_scripts`, `\AMP_Template_Customizer::add_legacy_customizer_scripts` and `\AMP_Template_Customizer::add_customizer_scripts` would need to execute the `amp_register_polyfills` action as well.

Thanks, I suspected there might be more 😄  Covered that and the action docblock in cc504ca53

pierlon committed:
Ensure JS vendor directory is included in build

pierlon committed:
Include JS maps if it's a development build

pierlon commented:
@googlebot I consent.

pierlon commented:
When the plugin is built, the `assets/vendor` directory was not being copied to the `build` folder, and thus not included in the zip. I've fixed that in https://github.com/ampproject/amp-wp/pull/5164/commits/366a5e5bd761fdf69388b832e8167de3e00fed71.

johnwatkins0 commented:
> When the plugin is built, the `assets/vendor` directory was not being copied to the `build` folder, and thus not included in the zip. I've fixed that in [366a5e5](https://github.com/ampproject/amp-wp/commit/366a5e5bd761fdf69388b832e8167de3e00fed71).

Thanks for the fix. That could also have something to do with the #5147 bug. Hard to say at this point 🤷‍♂️

pierlon commented:
Possibly. I tested the development build on WP 4.9 (before the fix) and the Settings page failed to load, but the Onboarding Wizard seemed fine (didn't test all the pages though).

westonruter commented:
> Possibly. I tested the development build on WP 4.9 (before the fix) and the Settings page failed to load, but the Onboarding Wizard seemed fine (didn't test all the pages though).

But after the changes here, the Settings Screen and the Onboarding Wizard both load as expected?

pierlon commented:
Yes both work fine now.

westonruter commented on Gruntfile.js:
```diff
@@ -54,7 +54,13 @@ module.exports = function( grunt ) {
 			compiled: {
-				src: [ 'assets/js/*.js', '!assets/js/amp-service-worker-runtime-precaching.js', 'assets/js/*.asset.php' ],
+				src: [
+					'assets/js/**/*.js',
+					'assets/js/**/*.js.map',
+					'!assets/js/amp-service-worker-runtime-precaching.js',
+					'assets/js/**/*.asset.php',
+					'assets/css/*.css',
```
Oh good. This should address an issue where running a build could accidentally include files that aren't on the branch, since they were leftover from builds on previous branches.

westonruter commented on Gruntfile.js:
```diff
@@ -155,11 +161,15 @@ module.exports = function( grunt ) {
 			paths.push( 'patches/*.patch' );
 
+			if ( 'development' === process.env.NODE_ENV ) {
+				paths.push( 'assets/js/**/*.js.map' );
```
Very nice.

westonruter commented on includes/class-amp-theme-support.php:
```diff
@@ -2298,6 +2298,11 @@ public static function setup_paired_browsing_client() {
 		}
 
+		/**
+		 * Fires before registering plugin assets that may require core asset polyfills.
```
```suggestion
		 * Fires before registering plugin assets that may require core asset polyfills.
		 *
		 * @internal
```

westonruter commented on src/Admin/Polyfills.php:
```diff
@@ -0,0 +1,140 @@
+	 * @return string Registration action to use.
+	 */
+	public static function get_registration_action() {
+		return 'amp_register_polyfills';
```
Why is this not using `wp_default_scripts` and `wp_default_styles`? 

Granted this would need to be not using `Delayed`, as the `register` method.

westonruter replied:
I suppose not, though, as the intention is for the polyfills to only be run on specific screens.

westonruter replied:
And `amp_register_default_scripts()` was doing this before?

johnwatkins0 replied:
Yes, `amp_register_default_scripts` was doing that before. With the new approach it will only happen when `do-action( 'amp_register_polyfills' )` is called, and this has been placed throughout the plugin where needed.

westonruter commented on src/Admin/Polyfills.php:
```diff
@@ -0,0 +1,140 @@
+	 * Runs on instantiation.
+	 */
+	public function register() {
+		$this->register_shimmed_scripts( wp_scripts() );
+		$this->register_shimmed_styles( wp_styles() );
```
Per above, then this could be done:

```suggestion
		add_action( 'wp_default_scripts', [ $this, 'register_shimmed_scripts' ] );
		add_action( 'wp_default_styles', [ $this, 'register_shimmed_styles' ] );
```

johnwatkins0 replied:
I did try this at first, but `wp_default_styles` runs too early for the contextual information we need to register the polyfills conditionally.

westonruter commented on tests/php/src/Admin/PolyfillsTest.php:
```diff
@@ -0,0 +1,83 @@
+		$this->instance->register();
+
+		$wp_scripts = new WP_Scripts();
+		$wp_scripts->init();
```
The `init` function is already called in the constructor.

```suggestion
```

westonruter commented:
> 2\. Created a `Polyfills` class that registers core assets that may not be available in earlier versions of WP. It hooks into the `wp_default_scripts` and `wp_default_styles` hooks immediately after the core defaults are registered. When the needed scripts/style are not registered by core, we register the versions provided by the plugin.

This part of the PR description doesn't seem current, is it?

johnwatkins0 committed:
Mark action as internal

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Remove unnecessary init call on wp_scripts instance

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Fix equals sign alignment

pierlon committed:
Reindex array keys to ensure they are sequential

This should fix an issue where converting the value to JSON produces an object instead of an array.

westonruter committed:
Allow legacy theme name to be translated, update description, use constant, s/classic/legacy/g
